pkg/apis/security/v1alpha1/types.go

```diff
@@ -45,8 +45,10 @@ type NetworkPolicySpec struct {
 	// Priority specfies the order of the NetworkPolicy relative to other
 	// NetworkPolicies.
 	Priority float64 `json:"priority"`
-	// Select workloads on which the rules will be applied to.
-	AppliedTo []NetworkPolicyPeer `json:"appliedTo"`
+	// Select workloads on which the rules will be applied to. Cannot be set in
+	// conjunction with AppliedTo in each rule.
+	// +optional
+	AppliedTo []NetworkPolicyPeer `json:"appliedTo,omitempty"`
```

Comment:
@tnqn do you know of any backwards-compatibility concern when adding `omitempty` to an existing field? I would imagine there isn't one.

Reply:
We didn't use these types to create the instances so I guess it won't affect anything. From some doc I searched, omitempty affects serialization, I'm not sure if it affects deserialization, like what we will get if the field is set to an empty slice when it's stored, nil or empty slice?
But what's the intention here to add omitempty? From the link @antoninbas gave, I saw that protobuf cannot tell the difference between nil and empty slice but json can, so should we avoid making them different meaning?

Reply:
Intension was to save the msg size. Before this change, `AppliedTo` in policy spec cannot be an empty slice, which I thought was the reason why no 'omitempty' was added. And we are not making an empty slice a different meaning than nil here... just want to make sure that no matter appliedTo is set in spec or rule, the other appliedTo will always be nil.

Reply:
Thanks for the explanation. Then will we have upgrade issue if a policy with empty AppliedTo was created before updating the CRD? Since kube-apiserver uses json to persist CRs, "appliedTo" will be present in etcd and we will get an empty slice instead of `nil` after deserializing it, would it cause problem in our process?
Should we use `len(appliedTo) == 0` since `nil` and empty slice mean the same.

Reply:
Sure, thanks for the analysis. Changed all nil checks to `len(appliedTo) == 0`.